New memory domain logic

jsarha opened the pull request:
Let's put this on hold until I get my experiment about implicit interim vs. lifetime heap division. E.g. Since the lifetime allocations are done at pipeline and module init phase and the interim allocations later - at runtime - it should be possible to create the interim heap when the first interim allocation arrives and use all memory that is left of lifetime heap. If this scheme works well, we can drop the interim vs. lifetime heap division.


The original description has been obsolete already for some time:

```
his is my first step towards implementing pipeline specific heaps. There is brand new create message payload definition and implementation. However, the target has already moved so that this payload is not enough anymore. 

There is also a problem that it looks like the FW receives some pipeline create messages with payload bit set, but the payload is not there. Firmware receives two messages where linux driver only logs about sending only one. This probably has something to do with forwarding the messages form core 0 to core 2. I have no idea how this can happen, but as I should move to work on kernel heap, I leave the debugging for now. 

The FW side code is here: https://github.com/thesofproject/sof/pull/10265
```

Commit:
ASoC: SOF: ipc4-topology: Add payload to pipeline create messages

Start adding payloads to pipeline create messages. The payload
contains information for payload specific memory configuration.

All non DP module instances within the same pipeline share the same
memory attributes and access the same resources. The new logic sums
interim, lifetime, and shared heap memory requirements together and
picks the highest stack requirement of all module instances belonging
to a pipeline. These pipeline specific attributes are sent as struct
sof_ipc4_glb_pipe_payload payload in pipeline's create message.

The idea is to pass common memory configuration for all the Low
Latency modules in the pipeline in pipeline create message payload.

The Data Processing module instances will still have an individual
memory configuration in struct sof_ipc4_mod_init_ext_dp_memory_data
payloads as before. In their payload everything is as it was before,
all attributes are copied directly from their topology attributes.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: SOF: ipc4-topology: Remove dp_ from all module memory attributes

Remove dp-prefix from all module instance's memory attributes and
related data structures. The attributes are not anymore exclusively
for Data Processing module instances, but generic for all module
instances. However, the module init payload is still only for DP
module instances.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

jsarha commented:
FYI @ranj063 @lyakh @lgirdwood

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1308,6 +1308,25 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+
+		if (swidget->domain_id != 0)
+			dev_warn(sdev->dev, "%s is not DP, but domain_id %u was set in topology",
+				 swidget->widget->name, swidget->domain_id);
```
if we've going to override it anyway, do we need a warning at all? Wont it be just noise?

jsarha replied:
This logic is about to change. We need separate domain_id and pipeline_id. So this code is already obsolete, but I don't try to reimplement it yet, since it seems the target is still moving.

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1308,6 +1308,25 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	if (swidget->comp_domain != SOF_COMP_DOMAIN_DP) {
+		pipe_widget->heap_bytes += swidget->heap_bytes;
+		if (pipe_widget->stack_bytes < swidget->stack_bytes)
+			pipe_widget->stack_bytes = swidget->stack_bytes;
```
use MAX?

lyakh replied:
@ranj063 I'm actually actively against using `MAX()` / `MIN()` in such cases. Using `a = MAX(a, b)` decodes to (leaving all the type-preservation and multiple evaluation avoidance off):
```
if (a < b)
    a = b;
else
    a = a;
```
where that latter `a = a` obviously makes no sense. Now, the compiler would hopefully drop it, but why should we write something "silly" to hope that the compiler rescues us? Only when the LHS isn't also one of the arguments `a = MAX(b, c)` then it makes perfect sense (to me). Just my personal opinion of course

Commit:
ASoC: ipc4: Add SOF_IPC4_GLB_CREATE_PIPELINE payload macros and structs

Adds SOF_IPC4_GLB_PIPE_EXT_OBJ_ARRAY macros to set extension bit in
SOF_IPC4_GLB_CREATE_PIPELINE indicating presence of the payload, and
all necessary macros and structs to create the payload.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
ASoC: sof: ipc4-topology: Fix SOF_TKN_COMP_STACK_BYTES_REQUIREMENT id

The was inconsistency with SOF_TKN_COMP_STACK_BYTES_REQUIREMENT and
SOF_TKN_COMP_HEAP_BYTES_REQUIREMENT token ids in the Linux driver code
with SOF FW topology code. This commit fixes the Linux side to match
tools/topology/topology2/include/common/tokens.conf

See https://github.com/thesofproject/sof/blob/788861804ed08485496e979dd9c467c1a21b30c5/tools/topology/topology2/include/common/tokens.conf#L30

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

lyakh commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -164,6 +164,8 @@ static const struct sof_topology_token comp_ext_tokens[] = {
 	{SOF_TKN_COMP_STACK_BYTES_REQUIREMENT, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
 		offsetof(struct snd_sof_widget, stack_bytes)},
+	{SOF_TKN_COMP_STATIC_BYTES_REQUIREMENT, SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+		offsetof(struct snd_sof_widget, static_bytes)},
```
to the commit message: is it heap size for initialisation only or for the entire life-time of a single instance (estimated, possibly dependent on configuration or use-case)

lyakh commented on include/sound/sof/ipc4/header.h:
```diff
@@ -691,6 +695,78 @@ struct sof_ipc4_mod_init_ext_dp_memory_data {
 	u32 heap_bytes;		/* stack size in bytes, 0 means default size */
 } __packed __aligned(4);
 
+
```
2 empty lines?

lyakh commented on include/sound/sof/ipc4/header.h:
```diff
@@ -691,6 +695,78 @@ struct sof_ipc4_mod_init_ext_dp_memory_data {
+#define SOF_IPC4_GLB_PIPE_EXT_OBJ_ARRAY(x)	((x) << SOF_IPC4_GLB_PIPE_EXT_OBJ_ARRAY_SHIFT)
+
+struct sof_ipc4_glb_pipe_payload {
+	u32 word0;
```
could you do bitfields instead e.g.
```
u32 payload : 24;
u32 array : 8;
```
? Or at least pick up a more informative name like `payload_array` and maybe use `__bitwise`?

jsarha replied:
This is Linux kernel style, at least in SOF driver, and in all other Linux drivers I've been working on. In SOF driver every other IPC message and payload structure is defined with macros, not with bitfields. I don't think I should diverge from this in PR either.

lyakh commented on include/sound/sof/ipc4/header.h:
```diff
@@ -691,6 +695,78 @@ struct sof_ipc4_mod_init_ext_dp_memory_data {
+#define SOF_IPC4_GLB_PIPE_EXT_OBJ_WORDS(x)	((x) << SOF_IPC4_GLB_PIPE_EXT_OBJ_WORDS_SHIFT)
+
+struct sof_ipc4_glb_pipe_ext_object {
+	u32 header;
```
could also be split into sub-fields

jsarha replied:
same as above

bardliao commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3068,6 +3088,57 @@ static int sof_ipc4_widget_setup_msg_payload(struct snd_sof_dev *sdev,
+
+	dev_dbg(sdev->dev, "payload word0 %#x domain_id %u stack_bytes %u heap_bytes %u",
+		 payload_hdr->word0, mem_data->domain_id, mem_data->stack_bytes,
+		 mem_data->heap_bytes);
```
The indentation seems be incorrect.

lgirdwood commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -108,8 +108,8 @@
-#define SOF_TKN_COMP_HEAP_BYTES_REQUIREMENT	420
-#define SOF_TKN_COMP_STACK_BYTES_REQUIREMENT	421
+#define SOF_TKN_COMP_STACK_BYTES_REQUIREMENT	420
+#define SOF_TKN_COMP_HEAP_BYTES_REQUIREMENT	421
```
Can we call this _INTERIM_HEAP_

jsarha replied:
I guess I can. These tokens  were already part of 2.14, but then again they were not really used for anything then.

lgirdwood replied:
Is INTERIM in a later commit ?
EDIT it is. Resolved.

lgirdwood commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -110,6 +110,8 @@
 #define SOF_TKN_COMP_DOMAIN_ID			419
 #define SOF_TKN_COMP_STACK_BYTES_REQUIREMENT	420
 #define SOF_TKN_COMP_HEAP_BYTES_REQUIREMENT	421
+#define SOF_TKN_COMP_LIFETIME_BYTES_REQUIREMENT	422
```
and SOF_TKN_COMP_LIFETIME_HEAP_BYTES_REQUIREMENT

Commit:
ASoC: SOF: ipc4-topology: Change struct sof_ipc4_mod_init_ext_dp_memory_data

Change struct sof_ipc4_mod_init_ext_dp_memory_data to what is required
for the SOF FW userspace DP processing. The earlier version of the
firmware (v2.14) did not use the contents of the struct for anything,
and if it receives a struct that is larger than the original, the
extra words are simply ignored, so there should not be any problem in
changing the struct. The following FW versions will expect larger
struct and ignore anything that is smaller.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

jsarha commented:
Change module init IPC payload as requested by @lgirdwood .

lgirdwood commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3068,6 +3088,58 @@ static int sof_ipc4_widget_setup_msg_payload(struct snd_sof_dev *sdev,
+	u32 *payload;
+	u32 ext_pos;
+
+	payload = kzalloc(sdev->ipc->max_payload_size, GFP_KERNEL);
```
we are doing multiple ROUND_UP()s below, do we need to factor this in to the size here ?

jsarha replied:
All the sizes are indicated in number of 32-bit words and the payload array is also indexed as uint32_t table. So all should be good. If there would be structs with sizes not multiple of sizeof(uint32_t) in the payload then those would be padded to up to the nearest uint32_t border. (In fact all the related structs have sizes that are multiple of  sizeof(uint32_t)). In the end the size of the whole payload is taken from the uint32_t table index so everything should be good.

lgirdwood commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3221,8 +3293,8 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
-		ret = sof_ipc4_widget_setup_msg_payload(sdev, swidget, msg, ipc_data, ipc_size,
-							&ext_data);
+		ret = sof_ipc4_widget_mod_init_msg_payload(sdev, swidget, msg, ipc_data, ipc_size,
+							   &ext_data);
```
are we freeing the payload on error ?

jsarha replied:
Yes, see line 3340.

lgirdwood reviewed (commented):
Just a few questions

kv2019i commented on include/sound/sof/ipc4/header.h:
```diff
@@ -690,9 +690,11 @@ enum sof_ipc4_mod_init_ext_obj_id {
+	uint32_t stack_bytes;		/* required stack size in bytes */
+	uint32_t interim_heap_bytes;	/* required interim heap size in bytes */
+	uint32_t lifetime_heap_bytes;	/* required lifetime heap size in bytes */
+	uint32_t shared_bytes;		/* required shared memory size in bytes */
```
Reviewing this together with https://github.com/thesofproject/sof/pull/10265 makes this a bit easier to follow. I still find relation of interim and lifetime a bit confusing, but additional notes in #10265 clarify this a bit (lifetime_heap_bytes memory needed in init and interim_heap_bytes stuff that may be allocated during runtime).

lyakh replied:
why change types? I think `u32` is preferred in the kernel

ujfalusi commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3234,6 +3307,17 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
+		if (ret > 0) {
+			ipc_size = ret;
+			ipc_data = ext_data;
+		}
```
Can you take a look at this patch in the compressed support series?
https://github.com/thesofproject/linux/pull/5647/commits/49d37d0e61388969b39e6dfb070a3f380181a4d9

Would it be possible to align with that implementation, I think it would be much easier to extend (adding the compr on top of this), that is.

lgirdwood replied:
ack - reuse would be best.

jsarha replied:
@ujfalusi Where can I find this commit (fetching your github or sof-linux did not work)? Easiest would be if I could rebase my branch on top of the compressed support code. Is it close to merging stage? Could it be spit so so that the mod init payload code would be merged early, before the rest?

jsarha replied:
Hmm, I am not any more sure what I should reuse, or to what exactly to align with.  

I could of course, just within this PR, write a common function to add either struct sof_ipc4_glb_pipe_ext_obj_memory_data or  struct sof_ipc4_mod_init_ext_dp_memory_data to a payload object array. The structs are actually identical, but by definition they are different structs an parts of different payloads, even if the object array concept is the same in the both payloads. That is not what you are asking, is it?

But if I could get the processing module configuration commit early and rebase this PR on top of that, then that should solve the merging problem with compress PR, shouldn't it?

ujfalusi replied:
Yes, that what I meant, to follow the pattern so it is going to be easier to move forward with other changes (with the way you do it is going to be a bit of a pain to rework the compr stuff, not that it is hard, just trying to cut corners)
The patch is `ASoC: SOF: ipc4-topology: Support init_ext_module_data for process modules` in https://github.com/thesofproject/linux/pull/5647

Commit:
ASoC: SOF: ipc4-topology: Refactor sof_ipc4_widget_mod_init_msg_payload()

Refactor sof_ipc4_widget_mod_init_msg_payload() to be easier to extend.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

lyakh commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3096,25 +3111,23 @@ static int sof_ipc4_widget_mod_init_msg_payload(struct snd_sof_dev *sdev,
+		 *       function.
+		 */
+		kfree(payload);
+		return 0;
```
nitpick: I'd do `if (!hdr) {kfree(payload); return 0;}` and then continue below with the rest

jsarha replied:
I wanted to have the both sides of the if statement, because it may be that sof_ipc4_add_init_ext_dp_memory_data() did not add  an object to the object array, and then we should not add the array at all, let alone mark the last element. There was an ext init payload even before my time and there are some bits defined there, so we may need to add the payload even if the array is not present. In that situation if (hdr) part remains, but the else branch should be removed. If I am to do something about this, then that would be adding the non array part always, e.g removing the the else branch.

lyakh replied:
sorry, my proposal wasn't really addressing the meaning of the code, it was rather cosmetic with no change to functionality. What this your hunk does is organise code as follows:
```
if (x) {
  a;
  b;
} else {
  y;
  return;
}
c;
d;
}
```
which means that `a`, `b`, `c` and `d` are always either executed or not executed together, so making some of them under the `if` and others outside of it is a bit strange. And that's exactly equivalent to
```
if (!x) {
  y;
  return;
}
a;
b;
c;
d;
}
```
which IMHO is more straight-forward. Also regarding your remark about removing the `else` part - the above would make it even easier by removing the whole of `if`.

lyakh commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1348,6 +1348,26 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	if (swidget->comp_domain != SOF_COMP_DOMAIN_DP) {
+		pipe_widget->interim_bytes += swidget->interim_bytes;
+		pipe_widget->lifetime_bytes += swidget->lifetime_bytes;
+		pipe_widget->shared_bytes += swidget->shared_bytes;
```
I suppose in theory these can overflow... We don't expect any realistic configurations to do that, but a malicious one could, especially once we start loading unsigned userspace modules

jsarha replied:
I do not think adding any safeguards to the Linux driver side serves any purpose. The FW however should have some sanity checks about the values give to it through IPC.

kv2019i commented on include/sound/sof/ipc4/header.h:
```diff
@@ -686,9 +690,83 @@ enum sof_ipc4_mod_init_ext_obj_id {
+	uint32_t stack_bytes;		/* required stack size in bytes */
+	uint32_t interim_heap_bytes;	/* required interim heap size in bytes */
+	uint32_t lifetime_heap_bytes;	/* required lifetime heap size in bytes */
+	uint32_t shared_bytes;		/* required shared memory size in bytes */
```
Ok this looks weird to mix "u32" and "uint32_t" in the same struct.

jsarha replied:
Ok.  Forgot my decision to got to u32 (despite the FW side headers using uint32_t) as its the Linux way. I'll update it.

kv2019i commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3105,55 +3129,70 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
-		return 0;
-	}
-
 	payload = kzalloc(sdev->ipc->max_payload_size, GFP_KERNEL);
 	if (!payload)
 		return -ENOMEM;
```
I think this is valid, the alloc is not needed in all cases.

jsarha replied:
Sure,  but as its on directly part of any commits in this series, I'll apply it as an additional optimization on top.

kv2019i reviewed (commented):
A few minor comments inline, but starting to look good.

Commit:
ASoC: SOF: ipc4-topology: Optimize sof_ipc4_widget_mod_init_msg_payload()

Optimize sof_ipc4_widget_mod_init_msg_payload() so that it skips the
payload allocation and rest of the function if no payload is needed.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

jsarha commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1344,6 +1348,26 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	 */
+	if (swidget->comp_domain != SOF_COMP_DOMAIN_DP) {
+		pipe_widget->interim_bytes += swidget->interim_bytes;
+		pipe_widget->lifetime_bytes += swidget->lifetime_bytes;
+		pipe_widget->shared_bytes += swidget->shared_bytes;
+		if (pipe_widget->stack_bytes < swidget->stack_bytes)
+			pipe_widget->stack_bytes = swidget->stack_bytes;
```
It looks like the Copilot does not have the slightest idea of what is going on here.

kv2019i replied:
Not just the copilot :) I assume pipeline object is built from scratch, so this reset logic is needed, but good to double-check. But I'm ok if you have checked for new pipelines we are starting from zero always (and we have no code-paths to refresh same widgets again).

jsarha commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -111,8 +111,10 @@
+#define SOF_TKN_COMP_STACK_BYTES_REQUIREMENT	420
+#define SOF_TKN_COMP_INTERIM_HEAP_BYTES_REQUIREMENT	421
+#define SOF_TKN_COMP_LIFETIME_HEAP_BYTES_REQUIREMENT	422
+#define SOF_TKN_COMP_SHARED_BYTES_REQUIREMENT	423
```
The topology tokens were not really used for anything yet on the firmware side so its safe change the definitions.

kv2019i commented on sound/soc/sof/sof-audio.h:
```diff
@@ -459,10 +459,12 @@ struct snd_sof_widget {
-	u32 dp_domain_id;	/* DP process userspace domain ID */
-	u32 dp_stack_bytes;	/* DP process stack size requirement in bytes */
-	u32 dp_heap_bytes;	/* DP process heap size requirement in bytes */
+	/* Module instance's memory configureation. */
```
This is still there. Not critical, but if you update anyways...

kv2019i reviewed (dismissed):
Looks good now. A few minor nits, but not worth it unless you update the series otherwise...

jsarha commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1369,6 +1369,23 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	 * requirement of all module instances belonging to the same
+	 * pipeline.
+	 */
+	if (swidget->comp_domain != SOF_COMP_DOMAIN_DP) {
```
Unset will never happen.

jsarha commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -1369,6 +1369,23 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
+	 * required heap sizes to the sum of all modules instance's
+	 * belonging to same pipeline and find the maximum stack
+	 * requirement of all module instances belonging to the same
+	 * pipeline.
```
fixed

jsarha commented on include/sound/sof/ipc4/header.h:
```diff
@@ -686,9 +690,79 @@ enum sof_ipc4_mod_init_ext_obj_id {
+ * Macros for creating struct sof_ipc4_glb_pipe_payload payload with
+ * its associated data. ext_init payload should be the first piece of
+ * payload following SOF_IPC4_GLB_CREATE_PIPELINE msg, and its
+ * existence is indicated with SOF_IPC4_GLB_PIPE_PAYLOAD bit.
```
fixed

jsarha commented on include/sound/sof/ipc4/header.h:
```diff
@@ -686,9 +690,79 @@ enum sof_ipc4_mod_init_ext_obj_id {
+	SOF_IPC4_GLB_PIPE_DATA_ID_MAX = SOF_IPC4_GLB_PIPE_DATA_ID_MEM_DATA,
+};
+
+/* Pipeline memory configuration data object for ext_init object array */
```
fixed

jsarha commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -111,8 +111,10 @@
+#define SOF_TKN_COMP_STACK_BYTES_REQUIREMENT	420
+#define SOF_TKN_COMP_INTERIM_HEAP_BYTES_REQUIREMENT	421
+#define SOF_TKN_COMP_LIFETIME_HEAP_BYTES_REQUIREMENT	422
+#define SOF_TKN_COMP_SHARED_BYTES_REQUIREMENT	423
```
This is explained in the commit message.

jsarha commented on include/sound/sof/ipc4/header.h:
```diff
@@ -686,9 +690,83 @@ enum sof_ipc4_mod_init_ext_obj_id {
+ * Macros for creating struct sof_ipc4_glb_pipe_payload payload with
+ * its associated data. ext_init payload should be the first piece of
+ * payload following SOF_IPC4_GLB_CREATE_PIPELINE msg, and its
+ * existence is indicated with SOF_IPC4_GLB_PIPE_PAYLOAD bit.
```
fixed

jsarha commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -3176,6 +3225,72 @@ static int sof_ipc4_widget_setup_msg_payload(struct snd_sof_dev *sdev,
+	/* Mark end of object array */
+	hdr->header |= SOF_IPC4_GLB_PIPE_EXT_OBJ_LAST_MASK;
+
+	/* Put total payload size in words to the payload header */
+	payload_hdr->word0 |= SOF_IPC4_GLB_PIPE_PAYLOAD_WORDS(ext_pos);
+	*new_data = payload;
+
+	/* Update msg extension bits according to the payload changes */
+	msg->extension |= SOF_IPC4_GLB_PIPE_PAYLOAD_MASK;
+
+	dev_dbg(sdev->dev, "%s: payload word0 %#x", swidget->widget->name,
+		payload_hdr->word0);
+
+	return ext_pos * sizeof(int32_t);
```
The max size of the payload is predefined, it can not increase depding on the input the function gets. Its simply impossible that the buffer would overflow.
